src/Microsoft.ML.Data/TrainCatalog.cs

```diff
@@ -61,14 +85,68 @@ public abstract class TrainCatalogBase
                 Complement = false
             }, data);
 
-            return (trainFilter, testFilter);
+            return new TrainTestData(trainFilter, testFilter);
+        }
+
+        /// <summary>
+        /// Results for specific cross validation fold.
+        /// </summary>
+        protected internal struct CrossValidationResult
+        {
+            /// <summary>
+            /// Model trained during cross validation fold.
+            /// </summary>
+            public readonly ITransformer Model;
+            /// <summary>
+            /// <see cref="IDataView"/> for scored fold.
+            /// </summary>
+            public readonly IDataView Scores;
+            /// <summary>
+            /// Fold number.
+            /// </summary>
+            public readonly int Fold;
+
+            public CrossValidationResult(ITransformer model, IDataView scores, int fold)
+            {
+                Model = model;
+                Scores = scores;
+                Fold = fold;
+            }
+        }
+
+        public class CrossValidationResult<T> where T : class
+        {
+            /// <summary>
+            /// Metrics for cross validation fold.
+            /// </summary>
+            public readonly T Metrics;
```

Comment:
>T [](start = 28, length = 1)

Part of me want to create interface and mark with it all Metric classes. Another part of me remembers what we try get rid of all empty interfaces.
Maybe empty base class?
Would like to hear your comments as well.

Reply:
+1 for some way of making this not generic.

---
In reply to: [255714622](https://github.com/dotnet/machinelearning/pull/2507#discussion_r255714622) [](ancestors = 255714622)